### PR TITLE
add pythonization of transformation

### PR DESCRIPTION
**Pull request by RTSandberg**
This PR exposes the coordinate transformation, allowing transformations ``to_fixed_s`` and ``to_fixed_t``.

- [x] [user-facing docs](https://github.com/ECP-WarpX/impactx/pull/360#issuecomment-1563510742)
- [x] test

### CODE REVIEW

File: src/python/transformation.cpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
+/* Copyright 2021-2023 The ImpactX Community
```
Comment (ax3l) [inline src/python/transformation.cpp:1]: @RTSandberg Don´t forget to add this `.cpp` file to the respective `CMakeLists.txt` :)

Comment (RTSandberg) [reply src/python/transformation.cpp:1]: Good catch :)

File: src/python/transformation.cpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * Authors: Ryan Sandberg, Axel Huebl
+ * License: BSD-3-Clause-LBNL
+ */
+#include "pyImpactX.H"
+
+#include <particles/ReferenceParticle.H>
+#include <particles/transformation/CoordinateTransformation.H>
+#include <AMReX.H>
```
Comment (ax3l) [inline src/python/transformation.cpp]: This include is unused :)
```suggestion
```

File: src/python/transformation.cpp [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace py = pybind11;
+using namespace impactx;
+
+void init_transformation(py::module& m)
+{
+    m.def("coordinate_transformation", &transformation::CoordinateTransformation,
+        "Transform coordinates."
+        // [](ImpactX)
```
Comment (ax3l) [inline src/python/transformation.cpp]:
```suggestion
    m.def("coordinate_transformation",
        &transformation::CoordinateTransformation,
        "Transform coordinates."
```

**Comment by ax3l** [comment]
To finalize, can you please add documentation for the new enum and freestanding function in https://impactx.readthedocs.io/en/latest/usage/python.html ?

You can add a new section `Coordinate Transformation`.

You can also add a test to cover this, e.g., in https://github.com/ECP-WarpX/impactx/tree/development/tests/python

File: docs/source/usage/python.rst [hunk 4]
```diff
@@ -637,3 +637,24 @@ This module provides elements for the accelerator lattice.
             (optional); default is a tanh fringe field model based on `<http://www.physics.umd.edu/dsat/docs/MaryLieMan.pdf>`__
    :param mapsteps: number of integration steps per slice used for map and reference particle push in applied fields
    :param nslice: number of slices used for the application of space charge
+
+Coordinate Transformation
+------------------------
```
Comment (ax3l) [inline docs/source/usage/python.rst]:
```suggestion
Coordinate Transformation
-------------------------
```

File: docs/source/usage/python.rst [hunk 5]
```diff
@@ -637,3 +637,24 @@ This module provides elements for the accelerator lattice.
...
+   :param to_fixed_s:
+
+Function
+.. py:method:: impactx.coordinate_transformation(pc, direction)
+
+   Function to transform the coordinates of the particles in a particle container either to fixed :math:`t` or to fixed :math:`s`.
+
+
+
+
```
Comment (ax3l) [inline docs/source/usage/python.rst]: Can you make this one single empty newline? :)

File: examples/transformation/REAMDE.rst [hunk 6]
```diff
@@ -0,0 +1,29 @@
+.. _examples-transformation:
+
+Transformation
+===================
```
Comment (ax3l) [inline examples/transformation/REAMDE.rst]:
```suggestion
Transformation
==============
```

File: examples/transformation/REAMDE.rst [hunk 7]
```diff
@@ -0,0 +1,29 @@
+.. _examples-transformation:
```
Comment (ax3l) [inline examples/transformation/REAMDE.rst]: Can you add this to `docs/source/usage/examples.rst`, maybe in a sub-section `Unit Tests` at the end?

Potential anchor name:
```suggestion
.. _tests-transformation:
```

File: examples/transformation/test_transformation.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    RefPart,
+    TransformationDirection,
+    coordinate_transformation,
+    distribution,
+    elements,
+)
+
+
+def get_particle_data(pc):
+    for lvl in range(pc.finest_level + 1):
```
Comment (ax3l) [inline examples/transformation/test_transformation.py]: Please add a doc string
```suggestion
    for lvl in range(pc.finest_level + 1):
        """Doc string here
        
        Does this copy the data?
        """
```

File: docs/source/usage/python.rst (hunk 5 above)
Comment (RTSandberg) [reply docs/source/usage/python.rst]: yep

File: examples/transformation/test_transformation.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+# Copyright 2022-2023 ImpactX contributors
+# Authors: Ryan Sandberg, Axel Huebl, Chad Mitchell
+# License: BSD-3-Clause-LBNL
+#
+# -*- coding: utf-8 -*-
+
+import numpy as np
+
+import amrex
+from impactx import (  # NOQA
```
Comment (ax3l) [inline examples/transformation/test_transformation.py]: Likely from prior experiment:
```suggestion
from impactx import (
```

File: docs/source/usage/examples.rst [hunk 10]
```diff
@@ -25,4 +25,12 @@ This section allows you to **download input files** that correspond to different
    examples/quadrupole_softedge/README.rst
    examples/positron_channel/README.rst
 
-For every change of the ImpactX code base, each of these examples are continuously tested and benchmarked.
+Unit tests
```
Comment (ax3l) [inline docs/source/usage/examples.rst:29]:
```suggestion

Unit tests
```

File: docs/source/usage/python.rst [hunk 11]
```diff
@@ -637,3 +637,21 @@ This module provides elements for the accelerator lattice.
             (optional); default is a tanh fringe field model based on `<http://www.physics.umd.edu/dsat/docs/MaryLieMan.pdf>`__
    :param mapsteps: number of integration steps per slice used for map and reference particle push in applied fields
    :param nslice: number of slices used for the application of space charge
+
+Coordinate Transformation
```
Comment (ax3l) [inline docs/source/usage/python.rst:642]:
```suggestion

Coordinate Transformation
```

**Review by ax3l — APPROVED** [review]
Awesome, thank you! :)

File: tests/python/test_transformation.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        sigmaPx=1.33 / energy_gamma,
+        sigmaPy=1.33 / energy_gamma,
+        sigmaPt=100 / energy_gamma,
+        muxpx=-0.5,
+        muypy=0.4,
+        mutpt=0.8,
+    )
+    sim.add_particles(bunch_charge_C, distr, npart)
+    # number of slices per ds in each lattice element
+    ns = 1
```
Comment (ax3l) [inline tests/python/test_transformation.py]:
```suggestion
```

**Pull request merged**